src/server/request_actor.rs

```diff
@@ -0,0 +1,365 @@
+                },
+                StatusCode::Cancelled,
+            );
+            // TODO
+        } else {
+            todo!()
+        }
+    }
+
+    fn send(&self, msg: RPCMessage) {
+        (self.sender)(msg);
+    }
+}
+
+struct CargoHandle {
+    /// The handle to the actual cargo process. As we cannot cancel directly from with
+    /// a read syscall dropping and therefore terminating the process is our best option.
+    child: GroupChild,
+    thread: jod_thread::JoinHandle<io::Result<(bool, String)>>,
+    receiver: Receiver<CargoMessage>,
+}
+
+impl CargoHandle {
+    fn spawn(mut command: Command) -> io::Result<CargoHandle> {
+        command
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .stdin(Stdio::null());
+        let mut child = command.group_spawn()?;
+
+        let stdout = child.inner().stdout.take().unwrap();
+        let stderr = child.inner().stderr.take().unwrap();
+
+        let (sender, receiver) = unbounded();
+        let actor = CargoActor::new(sender, stdout, stderr);
+        let thread = jod_thread::Builder::new()
+            .name("CargoHandle".to_owned())
+            .spawn(move || actor.run())
+            .expect("failed to spawn thread");
+        Ok(CargoHandle {
+            child,
+            thread,
+            receiver,
+        })
+    }
+
+    fn cancel(mut self) {
+        let _ = self.child.kill();
+        let _ = self.child.wait();
+    }
+
+    fn join(mut self) -> io::Result<()> {
+        let _ = self.child.kill();
+        let exit_status = self.child.wait()?;
+        let (read_at_least_one_message, error) = self.thread.join()?;
+        if read_at_least_one_message || exit_status.success() {
+            Ok(())
+        } else {
+            Err(io::Error::new(io::ErrorKind::Other, format!(
+                "Cargo watcher failed, the command produced no valid metadata (exit code: {:?}):\n{}",
+                exit_status, error
+            )))
+        }
+    }
+}
+
+struct CargoActor {
+    sender: Sender<CargoMessage>,
+    stdout: ChildStdout,
+    stderr: ChildStderr,
+}
+
+impl CargoActor {
+    fn new(sender: Sender<CargoMessage>, stdout: ChildStdout, stderr: ChildStderr) -> CargoActor {
+        CargoActor {
+            sender,
+            stdout,
+            stderr,
+        }
+    }
+
+    fn run(self) -> io::Result<(bool, String)> {
+        // We manually read a line at a time, instead of using serde's
+        // stream deserializers, because the deserializer cannot recover
+        // from an error, resulting in it getting stuck, because we try to
+        // be resilient against failures.
+        //
+        // Because cargo only outputs one JSON object per line, we can
+        // simply skip a line if it doesn't parse, which just ignores any
+        // erroneous output.
+        //
+        // We return bool that indicates whether we read at least one message and a string that
+        // contains the error output.
```

Comment:
Ale zamiast pisać to w komentarzu to proponuję wyciągnąć to do struktury z nazwanymi polami :)

Reply:
Jezeli uwazasz ze to konieczne to oczywiscie tak zrobie ale zmienna ktora zwracamy ma nazwe read_at_least_one_message, tak samo w momencie kiedy ją "odbieramy". Czy dodanie structa nie zaciemni tylko obrazu?